### PR TITLE
Ensure git directory is clean when running `theme pull`

### PR DESCRIPTION
**Pull request by lukeh-shopify**
### WHY are these changes introduced?

Another check to avoid a potentially destructive behaviour when running `theme pull`.

### WHAT is this pull request doing?

Checks that the current directory is a git directory, and that there are no uncommitted changes before pulling and overwriting new files.

### How to test your changes?

1. `git clone https://github.com/Shopify/dawn` or `git init` in a new directory
2. Touch a file in the directory
3. Run `shopify theme pull` within that directory
4. Ensure that you are prompted

https://github.com/user-attachments/assets/1e3dfd78-be60-48e5-a8ff-317f6662aecc

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes

### CODE REVIEW

File: packages/theme/src/cli/services/pull.ts [hunk 1]
```diff
@@ -13,6 +13,7 @@ import {fetchChecksums} from '@shopify/cli-kit/node/themes/api'
 import {renderSuccess} from '@shopify/cli-kit/node/ui'
 import {glob} from '@shopify/cli-kit/node/fs'
 import {cwd} from '@shopify/cli-kit/node/path'
+import {insideGitDirectory, isClean} from '@shopify/cli-kit/node/git'
```
Comment (graygilmore) [inline packages/theme/src/cli/services/pull.ts:16]: ~~Am I losing my mind or does `insideGitDirectory` not exist? I know we have `ensureInsideGitDirectory` but I can't see where we export insideGitDirectory.~~

It's added in this PR 🤦🏻‍♂️

File: packages/theme/src/cli/services/pull.ts [hunk 2]
```diff
@@ -149,6 +150,21 @@ async function executePull(theme: Theme, session: AdminSession, options: PullOpt
...
+  if (
+    (await insideGitDirectory(path)) &&
+    !(await isClean(path)) &&
+    !(await currentDirectoryConfirmed(
+      force,
+      'The current Git directory has uncommitted changes. Do you want to proceed?',
+    ))
+  ) {
+    return
+  }
```
Comment (graygilmore) [inline packages/theme/src/cli/services/pull.ts:166]: I think we could restructure this a tiny bit. There's no need to do any of the checks if `force` is `true`:

```suggestion
  if (!force) {
    const dirtyDirectory = (await insideGitDirectory(path)) && !(await isClean(path));
    if (
      dirtyDirectory && 
      !(await currentDirectoryConfirmed(
        force,
        'The current Git directory has uncommitted changes. Do you want to proceed?',
      ))
    ) {
      return
    }
```

Still a little awkward but maybe a touch clearer?

File: packages/theme/src/cli/services/pull.test.ts [hunk 3]
```diff
@@ -82,6 +87,25 @@ describe('pull', () => {
     )
   })
 
+  test('should ask for confirmation if the current directory is a Git directory and is not clean', async () => {
```
Comment (graygilmore) [inline packages/theme/src/cli/services/pull.test.ts:90]: Let's add a case for when `--force` is used to ensure the user isn't prompted.

**Review by graygilmore — COMMENTED** [review]
Good change!

**Comment by jeffreyguenther** [comment]
Nice polish!

File: packages/cli-kit/src/public/node/git.ts [hunk 4]
```diff
@@ -268,11 +268,23 @@ export class OutsideGitDirectoryError extends AbortError {}
 export async function ensureInsideGitDirectory(directory?: string): Promise<void> {
   // eslint-disable-next-line @typescript-eslint/ban-ts-comment
   // @ts-ignore
-  if (!(await git({baseDir: directory}).checkIsRepo())) {
+  if (!(await insideGitDirectory(directory))) {
     throw new OutsideGitDirectoryError(`${outputToken.path(directory || cwd())} is not a Git directory`)
   }
 }
 
+/**
```
Comment (jamesmengo) [inline packages/cli-kit/src/public/node/git.ts:276]: WDYT about moving this before the theme selection step? I'd want to know about this before selecting a theme to pull

Comment (lukeh-shopify) [reply packages/cli-kit/src/public/node/git.ts:276]: Happy to do that, was just following the existing convention for the invalid theme directory check!

I'll extract both checks in this PR

File: packages/theme/src/cli/services/pull.ts [hunk 5]
```diff
@@ -107,6 +108,10 @@ export async function pull(flags: PullFlags): Promise<void> {
 
   const {path, nodelete, live, development, only, ignore, force} = flags
 
+  if (!(await validateDirectory(path || cwd(), force || false))) {
+    return
```
[the remark this answers is not in the document]
Comment (lukeh-shopify) [reply packages/theme/src/cli/services/pull.ts:112]: Thanks, my typescript foo sucks :D

File: packages/theme/src/cli/services/pull.ts [hunk 6]
```diff
@@ -196,3 +184,46 @@ export async function isEmptyDir(path: string) {
...
+    !(await currentDirectoryConfirmed(force))
+  ) {
+    return false
+  }
+
+  /**
+   * If users are not forcing the 'pull' command, and the current directory is a
+   * Git directory and it is not clean, we ask for confirmation before proceeding.
+   */
+  if (!force) {
```
[the remark this answers is not in the document]
Comment (lukeh-shopify) [reply packages/theme/src/cli/services/pull.ts]: I like the early return approach, updated

**Review by isaacroldan — APPROVED** [review]

**Pull request merged**